### PR TITLE
Download image from Server

### PR DESCRIPTION
**Pull request by alknopfler**
Signed-off-by: Alberto Morgante Medina <alknopfler@users.noreply.github.com>

### CODE REVIEW

**Comment by alknopfler** [comment]
- adding ProgressJson to object
- Creating tests for progressJson function
- Splitting the doBootstrap function in two: Bootstrap Onboarding request, and download image
- Generating new function to downloadimage

File: sztp-agent/pkg/secureAgent/daemon.go [hunk 1]
```diff
@@ -103,8 +113,47 @@ func (a *Agent) doRequestBootstrapServer() error {
...
+}
+
+func (a *Agent) downloadAndValidateImage() error {
+	log.Printf("[INFO] Starting the Download Image: %v", a.BootstrapServerOnboardingInfo.IetfSztpConveyedInfoOnboardingInformation.BootImage.DownloadURI)
+	// Download the image from DownloadURI and save it to a file
+	a.BootstrapServerOnboardingInfo.IetfSztpConveyedInfoOnboardingInformation.InfoTimestampReference = fmt.Sprintf("%+8d", time.Now().Unix())
+	var wg sync.WaitGroup
+	wg.Add(len(a.BootstrapServerOnboardingInfo.IetfSztpConveyedInfoOnboardingInformation.BootImage.DownloadURI))
+	for i, item := range a.BootstrapServerOnboardingInfo.IetfSztpConveyedInfoOnboardingInformation.BootImage.DownloadURI {
+		//goroutine to make the download concurrent
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/daemon.go:128]: the download should not be concurrent here... 
I think the SPEC intent is if you fail to download first location, then move to second location.... 
if first image was downloaded and matched the SHA, you break and not continue to next URL...

File: sztp-agent/pkg/secureAgent/daemon.go [hunk 2]
```diff
@@ -52,16 +62,16 @@ func (a *Agent) getBootstrapURL() error {
 func (a *Agent) doReportProgress() error {
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/daemon.go:62]: doReportProgress should receive `ProgressType` argument

**Review by glimchb — COMMENTED** [review]
this is great progress, thanks

File: sztp-agent/pkg/secureAgent/daemon.go [hunk 3]
```diff
@@ -103,8 +113,47 @@ func (a *Agent) doRequestBootstrapServer() error {
...
+			//Create a empty file
+			file, err := os.Create(ARTIFACTS_PATH + prefix + name)
+			if err != nil {
+				return err
+			}
+
+			size, err := io.Copy(file, response.Body)
+			log.Printf("[INFO] Downloaded file: %s with size: %d", ARTIFACTS_PATH+prefix+name, size)
+			log.Println("[INFO] Verify the file checksum: ", ARTIFACTS_PATH+prefix+name)
+			//switch a.BootstrapServerOnboardingInfo.IetfSztpConveyedInfoOnboardingInformation.BootImage.ImageVerification[i].HashAlgorithm {
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/daemon.go:149]: in shell I use `openssl dgst -sha256 -c /tmp/${BASENAME}` see https://github.com/opiproject/sztp/blob/main/tests.sh#L65
you can repeat my step

**Review by glimchb — APPROVED** [review]
excellent

File: sztp-agent/main.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ func main() {
 	command := newCommand()
 	if err := command.Execute(); err != nil {
 		log.Fatalf(color.InRed("[ERROR]")+"%s", err.Error())
+		os.Exit(-1)
```
Comment (glimchb) [inline sztp-agent/main.go:23]: I think `Fatalf` already does `exit`

**Pull request merged**